include/pybind11/detail/internals.h

```diff
@@ -292,12 +292,16 @@ struct type_info {
 #endif
 
 /// On Linux/OSX, changes in __GXX_ABI_VERSION__ indicate ABI incompatibility.
-/// On MSVC, changes in _MSC_VER may indicate ABI incompatibility (#2898).
+/// On MSVC, mixing /MT and /MD will result in crashes. See (#4953)
 #ifndef PYBIND11_BUILD_ABI
 #    if defined(__GXX_ABI_VERSION)
 #        define PYBIND11_BUILD_ABI "_cxxabi" PYBIND11_TOSTRING(__GXX_ABI_VERSION)
-#    elif defined(_MSC_VER)
-#        define PYBIND11_BUILD_ABI "_mscver" PYBIND11_TOSTRING(_MSC_VER)
+#    elif defined(_MSC_VER) && defined(_MT)
```

Comment:
I don't know if the assumption that all `_MSC_VER >= 1900 && _MSC_VER < 2000` are fully binary compatible is valid, but I'm willing to take your word for it. I'll defer to @wjakob for make the final call on this one.

It would be ideal to reduce the redundancy and avoid the `define PYBIND11_BUILD_ABI ""` fallback. (IMO it would be good to get rid of that fallback entirely, but not in this PR.) How about this?

```
#    elif defined(_MSC_VER)
#        if defined(_MT)
#            define PYBIND11_BUILD_ABI "_mt_mscver" PYBIND11_TOSTRING(_MSC_VER)
#        elif defined(_MD)
#            if _MSC_VER >= 1900 && _MSC_VER < 2000
#                define PYBIND11_BUILD_ABI "_md_mscver14"
#            else
#                define PYBIND11_BUILD_ABI "_md_mscver" PYBIND11_TOSTRING(_MSC_VER)
#            else
#                define PYBIND11_BUILD_ABI "_unkown_mscver" PYBIND11_TOSTRING(_MSC_VER)
#            endif
#        endif
```

I'm not sure about the `_unkown_mscver` fallback. Maybe `#error Unkown combination of MSVC macros` (or similar) would be better?

Reply:
I'm using `_MSC_VER/100` now to get the major version. See https://learn.microsoft.com/en-us/cpp/overview/compiler-versions?view=msvc-170 on how _MSC_VER is formatted.

Reply:
I'll ask around to find someone familiar with the MSVC ABI universe.

It's important that we get this right this time around, because everytime we make a change here evidently we're creating a big problem for conda-forge.

Reply:
I forgot to ask, did you consider the idea of preserving `"_mscver"` for one branch of the `#if defined` logic?

https://github.com/pybind/pybind11/pull/4953#issuecomment-2206806268

The idea is to minimize ABI breakages.

I'm just not sure which branch is the best one to pick.